internal/controller/util.go

```diff
@@ -11,39 +34,55 @@ const (
 	FinalizerName = "k8s.ngrok.com/finalizer"
 )
 
+// IsUpsert returns true if the object is being created or updated. That is, if the deletion timestamp is not set.
 func IsUpsert(o client.Object) bool {
 	return o.GetDeletionTimestamp().IsZero()
 }
 
+// IsDelete returns true if the object is being deleted. That is, if the deletion timestamp is set and non-zero.
 func IsDelete(o client.Object) bool {
 	return !o.GetDeletionTimestamp().IsZero()
 }
 
+// HasFinalizer returns true if the object has the ngrok finalizer present.
+// It is our wrapper around controllerutil.ContainsFinalizer.
 func HasFinalizer(o client.Object) bool {
 	return controllerutil.ContainsFinalizer(o, FinalizerName)
 }
 
+// AddFinalizer accepts an Object and adds the ngrok finalizer if not already present.
+// It returns an indication of whether it updated the object's list of finalizers.
+// It is our wrapper around controllerutil.AddFinalizer.
 func AddFinalizer(o client.Object) bool {
 	return controllerutil.AddFinalizer(o, FinalizerName)
 }
 
+// RemoveFinalizer accepts an Object and removes the ngrok finalizer if present.
+// It returns an indication of whether it updated the object's list of finalizers.
+// It is our wrapper around controllerutil.RemoveFinalizer.
 func RemoveFinalizer(o client.Object) bool {
 	return controllerutil.RemoveFinalizer(o, FinalizerName)
 }
 
+// RegisterAndSyncFinalizer adds the ngrok finalizer to the object if not already present.
+// If it adds the finalizer, it updates the object in the Kubernetes API.
 func RegisterAndSyncFinalizer(ctx context.Context, c client.Writer, o client.Object) error {
-	if !HasFinalizer(o) {
-		AddFinalizer(o)
+	if AddFinalizer(o) {
 		return c.Update(ctx, o)
 	}
 	return nil
 }
 
+// RemoveAndSyncFinalizer removes the ngrok finalizer from the object if present.
+// If it removes the finalizer, it updates the object in the Kubernetes API.
 func RemoveAndSyncFinalizer(ctx context.Context, c client.Writer, o client.Object) error {
-	RemoveFinalizer(o)
-	return c.Update(ctx, o)
+	if RemoveFinalizer(o) {
```

Comment:
Same thing as above. We don't need to update the object unless the finalizer was removed. I think we could also eventually make this a patch instead of an update.